File: lib/aura/kernel/shadow_backup.rb
# frozen_string_literal: true

require "open3"
require "fileutils"

module Aura
  module Kernel
    class ShadowBackup
      MAX_FILE_SIZE = 1024 * 1024 # 1MB

      def initialize(project_path)
        @project_path = File.expand_path(project_path)
        @env_path = Aura::PathResolver.environment_path(@project_path)
        @shadow_path = File.join(@env_path, "shadow")
        @shadow_git = File.join(@shadow_path, ".git")
      end

      # Record any changed files in the project to the shadow workspace
      def record_changes(tool_name, tool_args = {})
        ensure_shadow_git_initialized!

        changed_files = []
        if File.exist?(File.join(@project_path, ".git"))
          # Use git status to find modified or untracked files
          out, _err, status = Open3.capture3("git", "status", "--porcelain", chdir: @project_path)
          if status.success?
            out.each_line do |line|
              # line format: " M path/to/file.py" or "?? path/to/file.py"
              filepath = line[3..].to_s.strip
              filepath = filepath.gsub(/\A"|"\z/, "") # strip quotes if git quotes the path
              changed_files << filepath
            end
          end
        else
          # Fallback: parse specific file writing arguments
          if tool_args.is_a?(Hash)
            file_path = tool_args["file_path"] || tool_args["path"]
            changed_files << file_path if file_path && !file_path.to_s.strip.empty?
          end
        end

        changed_files = changed_files.compact.uniq
        copied_any = false

        changed_files.each do |rel_path|
          abs_src = File.expand_path(rel_path, @project_path)

          # Ignore directories, non-existent files, and large files
          next unless File.file?(abs_src)
          next if File.size(abs_src) > MAX_FILE_SIZE

          # Prevent backing up the environment folder or files outside workspace
          next if rel_path.start_with?(".aura/") || rel_path.include?("/.aura/")

          begin
            real_src = File.realpath(abs_src)
            real_project = File.realpath(@project_path)
            next unless real_src.start_with?(real_project)
          rescue StandardError
            next
          end

          # Skip if ignored in the user's project git
          if File.exist?(File.join(@project_path, ".git"))
            _, _err, ignore_status = Open3.capture3("git", "check-ignore", rel_path, chdir: @project_path)
            next if ignore_status.success?
          end

          # Copy file into shadow path
          abs_dest = File.expand_path(rel_path, @shadow_path)
          FileUtils.mkdir_p(File.dirname(abs_dest))
          FileUtils.cp(abs_src, abs_dest)
          copied_any = true
        end

        if copied_any
          # Commit modifications into the shadow git repo
          message = "[Aura] Tool execution: #{tool_name}"
          Open3.capture3("git", "add", ".", chdir: @shadow_path)
          Open3.capture3("git", "commit", "-m", message, chdir: @shadow_path)
        end
      rescue StandardError => e
        # Fail-safe: don't break execution if backup error occurs
        warn "ShadowBackup Error: #{e.message}"
      end

      private

      def ensure_shadow_git_initialized!
        return if File.directory?(@shadow_git)

        FileUtils.mkdir_p(@shadow_path)
        Open3.capture3("git", "init", chdir: @shadow_path)
        Open3.capture3("git", "config", "user.name", "Aura Shadow Backup", chdir: @shadow_path)
        Open3.capture3("git", "config", "user.email", "shadow@aura-os.ai", chdir: @shadow_path)

        # Create initial commit
        gitignore_path = File.join(@shadow_path, ".gitignore")
        File.write(gitignore_path, "") unless File.exist?(gitignore_path)
        Open3.capture3("git", "add", ".gitignore", chdir: @shadow_path)
        Open3.capture3("git", "commit", "-m", "Initial commit", chdir: @shadow_path)
      end
    end
  end
end

File: src/core/kernel/shadowBackup.ts
import fs from 'node:fs';
import path from 'node:path';
import { execa } from 'execa';
import * as PathResolver from '../../utils/pathResolver.js';

export class ShadowBackup {
  public static readonly MAX_FILE_SIZE = 1024 * 1024; // 1MB

  private projectPath: string;
  private envPath: string;
  private shadowPath: string;
  private shadowGit: string;

  constructor(projectPath: string) {
    this.projectPath = path.resolve(projectPath);
    this.envPath =
      PathResolver.environmentPath(this.projectPath) || this.projectPath;
    this.shadowPath = path.join(this.envPath, 'shadow');
    this.shadowGit = path.join(this.shadowPath, '.git');
  }

  public async recordChanges(
    toolName: string,
    toolArgs: Record<string, unknown> = {},
  ): Promise<void> {
    try {
      await this.ensureShadowGitInitialized();

      let changedFiles: string[] = [];

      if (fs.existsSync(path.join(this.projectPath, '.git'))) {
        try {
          const { stdout } = await execa(
            'git',
            ['-c', 'core.quotepath=false', 'status', '--porcelain'],
            {
              cwd: this.projectPath,
            },
          );
          const lines = stdout.split('\n');
          for (const line of lines) {
            if (line.length > 3) {
              let filepath = line.substring(3).trim();
              if (filepath.startsWith('"') && filepath.endsWith('"')) {
                filepath = filepath.slice(1, -1);
              }
              changedFiles.push(filepath);
            }
          }
        } catch (_e) {}
      } else {
        if (toolArgs && typeof toolArgs === 'object') {
          const filePath = toolArgs.file_path || toolArgs.path;
          if (filePath && String(filePath).trim()) {
            changedFiles.push(String(filePath).trim());
          }
        }
      }

      changedFiles = Array.from(new Set(changedFiles.filter(Boolean)));
      let copiedAny = false;

      for (const relPath of changedFiles) {
        const absSrc = path.resolve(this.projectPath, relPath);
        const absDest = path.resolve(this.shadowPath, relPath);

        try {
          if (!fs.existsSync(absSrc)) {
            // Handle file deletion synchronization
            if (fs.existsSync(absDest)) {
              fs.rmSync(absDest, { force: true });
              copiedAny = true;
            }
            continue;
          }

          if (!fs.statSync(absSrc).isFile()) {
            continue;
          }

          if (fs.statSync(absSrc).size > ShadowBackup.MAX_FILE_SIZE) {
            continue;
          }

          if (relPath.startsWith('.aura/') || relPath.includes('/.aura/')) {
            continue;
          }

          // Security resolve check
          const realSrc = fs.realpathSync(absSrc);
          const realProject = fs.realpathSync(this.projectPath);
          if (!realSrc.startsWith(realProject)) {
            continue;
          }

          fs.mkdirSync(path.dirname(absDest), { recursive: true });
          fs.copyFileSync(absSrc, absDest);
          copiedAny = true;
        } catch (_e) {}
      }

      if (copiedAny) {
        const message = `[Aura] Tool execution: ${toolName}`;
        await execa('git', ['add', '.'], { cwd: this.shadowPath });
        await execa('git', ['commit', '-m', message], { cwd: this.shadowPath });
      }
    } catch (e: unknown) {
      const isClean =
        (e as Error).message &&
        ((e as Error).message.includes('nothing to commit') ||
          (e as Error).message.includes('working tree clean'));
      if (!isClean) {
        console.warn(`ShadowBackup Error: ${(e as Error).message}`);
      }
    }
  }

  private async ensureShadowGitInitialized(): Promise<void> {
    if (fs.existsSync(this.shadowGit)) {
      return;
    }

    fs.mkdirSync(this.shadowPath, { recursive: true });
    await execa('git', ['init'], { cwd: this.shadowPath });
    await execa('git', ['config', 'user.name', 'Aura Shadow Backup'], {
      cwd: this.shadowPath,
    });
    await execa('git', ['config', 'user.email', 'shadow@aura-os.ai'], {
      cwd: this.shadowPath,
    });

    const gitignorePath = path.join(this.shadowPath, '.gitignore');
    if (!fs.existsSync(gitignorePath)) {
      fs.writeFileSync(gitignorePath, '', 'utf-8');
    }

    await execa('git', ['add', '.gitignore'], { cwd: this.shadowPath });
    await execa('git', ['commit', '-m', 'Initial commit'], {
      cwd: this.shadowPath,
    });
  }
}
